File: src/service/DisciplineService.ts

import Discipline from "@/lib/models/Discipline";
import DisciplineRepository from "@/lib/repository/discipline/DisciplineRepository";
import UserRepository from "@/lib/repository/user/UserRepository";
import { message } from "@/types/TypeInterface";
import AuthStoreManager from "@/util/AuthStoreManager";

export default class DisciplineService {
    constructor(
        private disciplineRepository: DisciplineRepository,
        private userRepository: UserRepository,
    ) { }

    private async existyUser(): Promise<{ isbool: boolean, id: number | undefined }> {
        const result = await AuthStoreManager.get();
        if ( result && result.user) {
            return {
                isbool: await this.userRepository.existsById(result.user.id),
                id: result.user.id
            };
        }
        return {
            isbool: false,
            id: result?.user.id
        };
    }

    async createDisciplineService(
        name: string,
        description?: string
    ): Promise<message> {
        if (!name) {
            return {
                code: false,
                message: "Campos não preenchidos."
            }
        }
        const existsUser = await this.existyUser();
        if (!existsUser.isbool || !existsUser.id) {
            return {
                code: false,
                message: "Usuário não encontrado."
            }
        }

        if (await this.disciplineRepository.existsNameDiscipline(existsUser.id, name)){
            return {
                code: false,
                message: "Esta disciplina está cadastrada."
            }
        }

        const result = await this.disciplineRepository.createDiscipline(existsUser.id, name, description);
        if(!result){
            return {
                code: false,
                message: "Error ao gerar o id."
            }
        }
        return {
            code: true,
            message: "Disciplina criada com sucesso."
        };
    }

    async getAllDiscipline(): Promise<{ listDisc: Discipline[]; message: message }> {
        const existsUser = await this.existyUser();
        if (!existsUser.isbool || !existsUser.id) {
            return {
                listDisc: [],
                message: {
                    code: true,
                    message: "Usuário não encontrado"
                }
            }
        }

        const disciplines = await this.disciplineRepository.getAllDiscipline(existsUser.id);
        return {
            listDisc: disciplines,
            message: {
                code: true,
                message: "Disciplinas obtidas com sucesso"
            }
        };
    }

    async getDiscipline(id: number): Promise<{ listDisc: Discipline | null; message: message }> {
        const existsUser = await this.existyUser();
        if (!existsUser.isbool || !existsUser.id) {
            return {
                listDisc: null,
                message: {
                    code: true,
                    message: "Usuário não encontrado"
                }
            }
        }

        if (!await this.disciplineRepository.existsById(existsUser.id, id)) {
            return {
                listDisc: null,
                message: {
                    code: false,
                    message: "Disciplina não encontrado"
                }
            }
        }

        const discipline = await this.disciplineRepository.getDiscipline(existsUser.id, id);
        return {
            listDisc: discipline,
            message: {
                code: true,
                message: "Disciplina obtida com sucesso"
            }
        };
    }
}

File: src-tauri/src/service/discipline_service.rs
use crate::{
    error::app_error::AppError,
    model::{
        discipline::{UpdateDiscipline}
    },
    repository::{
        discipline_repository::DisciplineRepository,
        user_repository::{UserRepository},
    },
    service::dto::{discipline_response::DisciplineResponse, message_response::Message},
};

#[derive(serde::Serialize)]
pub struct DisciplineDataOne {
    pub message: Message,
    pub discipline: Option<DisciplineResponse>,
}

pub struct DisciplineDataAll {
    pub message: Message,
    pub discipline: Option<Vec<DisciplineResponse>>,
}

pub struct DisciplineService<'a> {
    discipline_repository: DisciplineRepository<'a>,
    user_repository: UserRepository<'a>,
}

impl<'a> DisciplineService<'a> {
    pub fn new(
        discipline_repository: DisciplineRepository<'a>,
        user_repository: UserRepository<'a>,
    ) -> Self {
        Self {
            discipline_repository,
            user_repository,
        }
    }

    pub async fn create_discipline(
        &self,
        user_id: i64,
        name: String,
        description: Option<String>,
    ) -> Result<Message, AppError> {
        if name.is_empty() {
            return Ok(Message {
                code: false,
                message: "Campos não preenchidos.".into(),
            });
        }

        if !self.user_repository.exists_by_id(user_id).await? {
            return Ok(Message {
                code: false,
                message: "O usuário não foi encontrado".into(),
            });
        }

        if !self
            .discipline_repository
            .exists_name_discipline(user_id, name.clone())
            .await?
        {
            return Ok(Message {
                code: false,
                message: "O nome de disciplina já existe, tente um outro nome".into(),
            });
        }

        self.discipline_repository
            .create_discipline(user_id, name, description)
            .await?;

        Ok(Message {
            code: true,
            message: "Disciplina criada com sucesso.".into(),
        })
    }

    pub async fn update_discipline(
        &self,
        user_id: i64,
        discipline_id: i64,
        name: Option<String>,
        description: Option<String>,
    ) -> Result<Message, AppError> {
        if !self.user_repository.exists_by_id(user_id).await? {
            return Ok(Message {
                code: false,
                message: "O usuário não foi encontrado".into(),
            });
        }

        if !self
            .discipline_repository
            .exists_by_id(discipline_id, user_id)
            .await?
        {
            return Ok(Message {
                code: false,
                message: "Disciplina não encontrada".into(),
            });
        }

        let update = UpdateDiscipline::new(name, description);
        self.discipline_repository
            .update_discipline(user_id, discipline_id, update)
            .await?;

        Ok(Message {
            code: true,
            message: "A disciplina foi atualizada com sucesso.".into(),
        })
    }

    pub async fn delete_discipline(
        &self, 
        user_id: i64,
        discipline_id: i64,
    ) -> Result<Message, AppError> {
        if !self.user_repository.exists_by_id(user_id).await? {
            return Ok(Message { 
                code: false, 
                message: "O usuário não foi encontrado".into() 
            });
        }

        if !self
            .discipline_repository
            .exists_by_id(discipline_id, user_id)
            .await?
        {
            return Ok(Message {
                code: false,
                message: "Disciplina não encontrada".into(),
            });
        }

        self.discipline_repository.delete_discipline(user_id, discipline_id).await?;

        Ok(Message { 
            code: true, 
            message: "A disciplina foi deletada".into() 
        })
    }

    pub async fn get_all_discipline(&self, user_id: i64) -> Result<DisciplineDataAll, AppError> {
        if !self.user_repository.exists_by_id(user_id).await? {
            return Ok(DisciplineDataAll {
                message: Message {
                    code: false,
                    message: "O usuário não foi encontrado".into(),
                },
                discipline: None,
            });
        }

        let result = self
            .discipline_repository
            .get_all_discipline(user_id)
            .await?;

        let discipline: Vec<DisciplineResponse> = result.iter().map(|d| d.into()).collect();

        Ok(DisciplineDataAll {
            message: Message {
                code: true,
                message: "Buscas efetuadas com sucesso".into(),
            },
            discipline: Some(discipline),
        })
    }

    pub async fn get_discipline(
        &self,
        user_id: i64,
        discipline_id: i64,
    ) -> Result<DisciplineDataOne, AppError> {
        if !self.user_repository.exists_by_id(user_id).await? {
            return Ok(DisciplineDataOne {
                message: Message {
                    code: false,
                    message: "O usuário não foi encontrado".into(),
                },
                discipline: None,
            });
        }

        if !self
            .discipline_repository
            .exists_by_id(discipline_id, user_id)
            .await?
        {
            return Ok(DisciplineDataOne {
                message: Message {
                    code: false,
                    message: "Disciplina não encontrada".into(),
                },
                discipline: None,
            });
        }

        let result = self.discipline_repository
            .get_discipline(discipline_id, user_id).await?;
        if result.is_none() {
            return Ok(DisciplineDataOne {
                message: Message {
                    code: false,
                    message: "Disciplina não foi encontrada".into(),
                },
                discipline: None,
            });
        }

        let discipline = DisciplineResponse::from(&result.unwrap());

        Ok(DisciplineDataOne {
            message: Message {
                code: true,
                message: "Busca efetuada com sucesso".into(),
            },
            discipline: Some(discipline),
        })
    }
}
